### PR TITLE
Added additional CRUD fields #208

### PR DESCRIPTION
**Pull request by sadhyama**

### CODE REVIEW

File: src/upstream.c [hunk 1]
```diff
@@ -389,3 +460,21 @@ void sendUpstreamMsgToServer(void **resp_bytes, size_t resp_size)
 	}
 
 }
+
+/*
+	Internal function to parse wrp src/dest
+*/
+static char *get_src_dest_from_sub_req(char *upstreamDest)
```
Comment (shilpa24balaji) [inline src/upstream.c]: Use wrp dest parsing function

Comment (sadhyama) [reply src/upstream.c]: We cannot use this wrp function here as this function is only for dest parsing.
For cloud-status retrieve requests, in addition to dest parsing we need to do source parsing as well in the reverse flow. i.e. while sending crud response to libparodus clients.
As it is not advised to use strtok, currently I have modified existing get_src_dest_from_req() with strchr and avoid strtok.

File: src/conn_interface.c [hunk 2]
```diff
@@ -150,9 +150,23 @@ void createSocketConnection(void (* initKeypress)())
             ParodusInfo("close_retry is %d, hence closing the connection and retrying\n", close_retry);
             close_and_unref_connection(get_global_conn());
             set_global_conn(NULL);
+
+            get_parodus_cfg()->cloud_status = "offline";
+            ParodusInfo("cloud_status set as %s after connection close\n", get_parodus_cfg()->cloud_status);
+            if(get_parodus_cfg()->cloud_disconnect !=NULL)
+            {
+				ParodusPrint("get_parodus_cfg()->cloud_disconnect is %s\n", get_parodus_cfg()->cloud_disconnect);
+				set_cloud_disconnect_time(5);
```
[the remark this answers is not in the document]
Comment (sadhyama) [reply src/conn_interface.c]: Incorporated.

File: src/crud_internal.c [hunk 3]
```diff
@@ -1188,12 +1210,119 @@ int updateObject( wrp_msg_t *reqMsg, wrp_msg_t **response )
...
-			return -1;
+			//Checks for parodus/cloud-disconnect request with objlevel = 3
+			if(objlevel == 3 && ((obj[3] != NULL) && (strcmp(obj[3] ,  "cloud-disconnect") == 0)))
+			{
+				if(reqMsg->u.crud.payload != NULL)
+				{
+					jsonPayload = cJSON_Parse( reqMsg->u.crud.payload );
+					if(jsonPayload !=NULL)
+					{
+						if((cJSON_GetObjectItem( jsonPayload, "disconnection-reason" )) !=NULL)
```
[the remark this answers is not in the document]
Comment (sadhyama) [reply src/crud_internal.c]: Incorporated.

File: src/crud_internal.c [hunk 4]
```diff
@@ -1188,12 +1210,119 @@ int updateObject( wrp_msg_t *reqMsg, wrp_msg_t **response )
...
+									cJSON_Delete(json);
+									json = NULL;
+									freeObjArray(&obj, objlevel);
+									return -1;
+								}
+								else
+								{
+									//check disconnection reason is character string of [a-zA-Z0-9 ]*
+
+									valid = validateDisconnectString(cJSON_GetObjectItem( jsonPayload,"disconnection-reason" )->valuestring);
```
[the remark this answers is not in the document]
Comment (sadhyama) [reply src/crud_internal.c]: Incorporated.

File: src/crud_internal.c [hunk 5]
```diff
@@ -1188,12 +1210,119 @@ int updateObject( wrp_msg_t *reqMsg, wrp_msg_t **response )
...
+									if(valid >0)
+									{
+										//set the disconnection-reason value to in-memory
+										get_parodus_cfg()->cloud_disconnect = strdup(cJSON_GetObjectItem( jsonPayload,"disconnection-reason" )->valuestring);
```
[the remark this answers is not in the document]
Comment (sadhyama) [reply src/crud_internal.c]: Incorporated.

File: src/config.c [hunk 6]
```diff
@@ -643,6 +645,8 @@ void setDefaultValuesToCfg(ParodusCfg *cfg)
     ParodusPrint("cfg->webpa_uuid is :%s\n", cfg->webpa_uuid);
     cfg->crud_config_file = strdup("parodus_cfg.json");
 	ParodusPrint("Default crud_config_file is %s\n", cfg->crud_config_file);
+	cfg->cloud_status = "offline";
```
[the remark this answers is not in the document]
Comment (sadhyama) [reply src/config.c]: Incorporated.

File: src/upstream.c (hunk 1 above)
Comment (sadhyama) [reply src/upstream.c]: As per the review suggestion, modified wrp-c parsing function to support both source and dest parsing and submitted PR in wrp-c. It is under review.

**Pull request merged**